.pre-commit-config.yaml

```diff
@@ -1,15 +1,10 @@
 repos:
-  - repo: https://github.com/psf/black
-    #this version is synced with the black mentioned in .github/workflows/ci.yml
-    rev: 23.10.0
+  - repo: https://github.com/astral-sh/ruff-pre-commit
+    rev: v0.2.1
     hooks:
-      - id: black
-        entry: bash -c 'black "$@"; git add -u' --
-        # It is recommended to specify the latest version of Python
-        # supported by your project here, or alternatively use
-        # pre-commit's default_language_version, see
-        # https://pre-commit.com/#top_level-default_language_version
-        language_version: python3.9
+      - id: ruff
+        args: [--fix]
```

Comment:
Ah here's the philosophical rub. We've talked about how appropriate it is to automatically fix bad code for the developer in a way that they aren't brought into the loop so they can learn to be better.

This does bring them into the loop a little bit, by not auto-committing the results at least. So I'd like to keep it that way, if/when we make the auto-formatting auto-committed.

I'd personally prefer to keep such checks out of the commit hooks altogether because of my WIP-commit-heavy workflow. I assume your workflow is different.

Anyway, we don't need to agree on this kind of thing. Just raising the point again.

Reply:
ok, per this and our chat off PR about this last night, I did some digging and found that pre-commit supports hooks at basically every lifecycle stage, so i've moved this check to pre-push. It didn't look like it is installed normally, so you have to specify the stage with `pre-commit install --hook-type pre-push`, but it then does the linting only on push to remote.

Reply:
Interesting... OK - we have the `pre-commit` instructions in our CONTRIBUTING.md docs usually (I think that's what we named them). - maybe those need updating.

It's annoying that pre-commit isn't more turnkey, but fine :smile:

Reply:
ah yep, meant to do that and spaced out - bumped the docs.